### PR TITLE
#642 Speeding up Inspect API's Data Structure Definition SPARQL Query

### PR DESCRIPTION
**Pull request by robons**
Initial informal tests suggest an approx 25% improvement in performance.

@nimshi89 and myself will be doing more thorough testing of this using the stress test, and then review profiler results - but it'd be good to get a review going on this ASAP.

Done as part of #642

### CODE REVIEW

File: src/csvcubed/cli/inspect/inspectdatasetmanager.py [hunk 1]
```diff
@@ -99,13 +101,15 @@ def get_standard_shape_measure_col_name_from_dsd(
         ComponentPropertyAttributeURI.MeasureType.value,
     )
 
-    if len(filtered_components) == 0 or filtered_components[0].csv_col_title == "":
+    if len(filtered_components) == 0 or filtered_components[0].used_in_columns == "":
```
Comment (robons) [inline src/csvcubed/cli/inspect/inspectdatasetmanager.py:104]: This is wrong.

File: src/csvcubed/cli/inspect/inspectdatasetmanager.py [hunk 2]
```diff
@@ -124,13 +128,15 @@ def get_standard_shape_unit_col_name_from_dsd(
         ComponentPropertyAttributeURI.UnitMeasure.value,
     )
 
-    if len(filtered_components) == 0 or filtered_components[0].csv_col_title == "":
+    if len(filtered_components) == 0 or filtered_components[0].used_in_columns == "":
```
Comment (robons) [inline src/csvcubed/cli/inspect/inspectdatasetmanager.py:131]: This won't work.

File: src/csvcubed/models/sparqlresults.py [hunk 3]
```diff
@@ -438,75 +449,77 @@ def _map_qube_component_sparql_result(
...
+        csv_column_definitions = map_csv_url_to_column_definitions[csv_url]
+
+        measure_uris = {
+            c.property for c in components if c.property_type == str(QB.MeasureProperty)
+        }
+
+        observed_value_columns = [
+            c
+            for c in csv_column_definitions
+            if (not c.virtual) and c.property_url in measure_uris
```
Comment (robons) [inline src/csvcubed/models/sparqlresults.py:490]: I don't think this matching is working at the moment. When debugging, I saw he `observed_value_columns` empty, when it needs to have some values. I think we need to do `startsWith` checking or similar.

File: src/csvcubed/models/sparqlresults.py [hunk 4]
```diff
@@ -438,75 +449,77 @@ def _map_qube_component_sparql_result(
...
             ]
 
-            if len(obs_val_col_title_result_for_component) == 1:
-                dsd_component_result.observation_value_column_titles = (
-                    obs_val_col_title_result_for_component[
-                        0
-                    ].observation_value_column_titles
-                )
+            component.used_in_columns = [
+                c for c in columns_using_this_component if c.title is not None
```
Comment (robons) [inline src/csvcubed/models/sparqlresults.py:501]: We should get rid of the `if c.title is not None` off the end here.

File: src/csvcubed/models/sparqlresults.py [hunk 5]
```diff
@@ -438,75 +449,77 @@ def _map_qube_component_sparql_result(
...
+                c.about_url
+                for c in columns_using_this_component
+                if c.about_url is not None
+            }
+
+            component.used_by_observed_value_columns = [
+                c
+                for c in observed_value_columns
+                if c.about_url in columns_using_this_component_about_urls
+                and c.title is not None
```
Comment (robons) [inline src/csvcubed/models/sparqlresults.py:514]: We should get rid of the `and c.title is not None` off the end here.

File: src/csvcubed/utils/csvdataset.py [hunk 6]
```diff
@@ -244,7 +248,7 @@ def transform_dataset_to_canonical_shape(
         obs_val_col_titles_about_urls = (
             data_cube_state.get_obs_val_col_titles_about_urls_for_csv(csv_url)
         )
-        col_names_col_titles = data_cube_state.get_col_name_col_title_for_csv(csv_url)
+        col_names_col_titles = data_cube_state.get_column_definitions_for_csv(csv_url)
```
Comment (robons) [inline src/csvcubed/utils/csvdataset.py:251]: We need to rename this variable.

File: tests/unit/utils/test_csvdataset.py [hunk 7]
```diff
@@ -419,13 +419,13 @@
 ]
 
 _col_names_col_titles = [
```
Comment (robons) [inline tests/unit/utils/test_csvdataset.py:421]: I'm sure this'll need to change quite a bit.

File: src/csvcubed/models/sparqlresults.py [hunk 8]
```diff
@@ -438,75 +449,77 @@ def _map_qube_component_sparql_result(
         property_type=get_component_property_type(
             str(result_dict["componentPropertyType"])
         ),
-        csv_col_title=none_or_map(result_dict.get("csvColumnTitle"), str) or "",
-        observation_value_column_titles="",  # This value is popuated after runnning the SELECT_OBS_VAL_FOR_DSD_COMPONENT_PROPERTIES sparql query. The _map_obs_val_for_dsd_component_properties_results will set the value for this property.
+        used_in_columns=none_or_map(result_dict.get("csvColumnTitle"), str) or "",
```
Comment (robons) [inline src/csvcubed/models/sparqlresults.py:452]: This is wrong.

File: src/csvcubed/cli/inspect/inspectdatasetmanager.py (hunk 1 above)
Comment (nimshi89) [reply src/csvcubed/cli/inspect/inspectdatasetmanager.py]: that means on line 108 "return filtered_components[0].csv_col_title" has to be changed as well right?

**Review by GDonRanasinghe — APPROVED** [review]
This looks good to me Rob - nice job!

**Pull request merged**